tpp-run/tpp-run.cpp

```diff
@@ -0,0 +1,281 @@
+//===- tpp-run.cpp - TPP CPU Execution Driver------------------------------===//
+//
+// Main entry point to a command line utility that executes an MLIR file on the
+// CPU by translating MLIR to LLVM IR before JIT-compiling and executing the
+// latter. Handles TPP/LIBXSMM include/library paths as well as benchmarking
+// modes, with warmups, measurements, output comparison, etc.
+//
+//===----------------------------------------------------------------------===//
+
+#include <iostream>
+
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/TargetSelect.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Transforms/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/ExecutionEngine/JitRunner.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Target/LLVMIR/Dialect/All.h"
+#include "mlir/InitAllDialects.h"
+#include "mlir/InitAllPasses.h"
+#include "mlir/Target/LLVMIR/ModuleTranslation.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+
+using namespace mlir;
+
+// This is a hack, to parse the command-line options locally
+// FIXME: Find a way to grab the options from the JitRunner
+struct CmdLineOpts {
+  std::string mainFuncName;
+} options;
+
+static void locaCmdLineOptParsing(int argc, char **argv) {
+  bool nextIsMain = false;
+  for (int i=0; i<argc; i++) {
+    if (strncmp("-e", argv[i], 2) == 0)
+      nextIsMain = true;
+    else if (nextIsMain) {
+      options.mainFuncName = argv[i];
+      break;
+    }
+  }
+}
+
+// Lowers to LLVM Dialect
+static LogicalResult lowerToLLVMDialect(ModuleOp module) {
+  // Minimal passes to make it work
+  // We don't want TPP passes here, as that's the job of tpp-opt
+  // The IR here should be free of TPP/XSMM or any TPP extensions
+  PassManager passManager(module.getContext());
+  applyPassManagerCLOptions(passManager);
+
+  // Bufferization, if needed
+  passManager.addNestedPass<func::FuncOp>(createTensorBufferizePass());
```

Comment:
We should prefer one shot bufferization instead of partial ones.

Reply:
Bufferization should happen *before* TPP passes, so these here are _just_ for the code that we generate in the wrapper call, and it doesn't need to be (and doesn't quite work when it is) global.

Reply:
mm, let's see. As long as we do not introduce extra allocs is fine.

Reply:
Even if it does, it's ok, because the kernel code would have been bufferized *before* this pass is done. These passes *only* bufferize the wrapper code, which isn't in the critical path and will not be in the timed portion, so we don't really care what happens there.